src/Features/Core/Portable/CustomMessageHandler/CustomMessageHandlerService.cs

```diff
@@ -0,0 +1,337 @@
+
+    /// <summary>
+    /// Handlers of non-document-related messages, indexed by handler message name.
+    /// </summary>
+    private readonly Dictionary<string, ICustomMessageHandlerWrapper> _handlers = new();
+
+    private readonly ICustomMessageHandlerFactory _customMessageHandlerFactory;
+
+    // Used to protect access to _extensions, _handlers and _documentHandlers.
+    private readonly object _lockObject = new();
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public CustomMessageHandlerService(ICustomMessageHandlerFactory customMessageHandlerFactory)
+    {
+        _customMessageHandlerFactory = customMessageHandlerFactory;
+    }
+
+    public ValueTask<RegisterHandlersResponse> LoadCustomMessageHandlersAsync(
+        Solution solution,
+        string assemblyFolderPath,
+        string assemblyFileName,
+        CancellationToken cancellationToken)
+    {
+        if (assemblyFileName.IndexOfAny(Path.GetInvalidFileNameChars()) >= 0)
+        {
+            throw new ArgumentException($"{assemblyFileName} is not a valid file name", nameof(assemblyFileName));
+        }
+
+        var analyzerAssemblyLoaderProvider = solution.Services.GetRequiredService<IAnalyzerAssemblyLoaderProvider>();
+        var assemblyPath = Path.Combine(assemblyFolderPath, assemblyFileName);
+
+        CustomMessageHandlerExtension? extension;
+        lock (_lockObject)
+        {
+            // Check if the assembly is already loaded.
+            if (!_extensions.TryGetValue(assemblyFolderPath, out extension))
+            {
+                var analyzerAssemblyLoader = analyzerAssemblyLoaderProvider.CreateNewShadowCopyLoader();
+
+                // Allow this assembly loader to load any dll in assemblyFolderPath.
+                foreach (var dll in Directory.EnumerateFiles(assemblyFolderPath, "*.dll"))
+                {
+                    try
+                    {
+                        // Check if the file is a valid .NET assembly.
+                        AssemblyName.GetAssemblyName(dll);
+                    }
+                    catch
+                    {
+                        // The file is not a valid .NET assembly, skip it.
+                        continue;
+                    }
+
+                    analyzerAssemblyLoader.AddDependencyLocation(dll);
+                }
+
+                extension = new CustomMessageHandlerExtension(analyzerAssemblyLoader);
+                _extensions[assemblyFolderPath] = extension;
+            }
+        }
+
+        // AssemblyLoadLockObject is only used to avoid multiple calls from the same extensions to load the same assembly concurrently
+        // resulting in the constructors of the same handlers being called more than once.
+        // All other concurrent operations, including modifying extension.Assemblies are protected by _lockObject.
+        lock (extension.AssemblyLoadLockObject)
```

Comment:
there is an alternative path I should have brought up before.  It is possible to mark the `CustomMessageLoadHandler` with `MutatesSolutionState => true` (and same for the unload).  this means we would never get parallel requests to load and unload the handlers, and any load/unload has to complete before the next request comes.

However it does have some considerations
1.  The implementation cannot throw, as throwing will cause the server to shut down.  For failures, we'd have to return a custom failure type instead.
2.   It is still possible for previously running requests (e.g. handle custom message) to still be running when a load/unload comes in.

I'm not sure it will be a huge amount cleaner though, so not blocking on this.